docs/proposals/14-wireguard-based-cluster-vpn.md

```diff
@@ -0,0 +1,178 @@
+
+On the seed, we add a WireGuard device to one node on the host network. For all other nodes on the seed, we adapt the routes accordingly to route traffic destined for the WireGuard network to our WireGuard node. The Kubernetes pods managing the WireGuard device and routes are only used for initial configuration and later reconfiguration. During runtime, they can restart without any impact on the operation of the WireGuard network as the WireGuard device is managed by the Linux kernel.
+
+The WireGuard endpoint on the seed is exposed via a load balancer. We propose to use kubelink to manage the WireGuard configuration/device on the seed. We consider the management of the WireGuard endpoint to be complex especially in error situations which is the reason for utilizing kubelink as there is already significant experience managing an endpoint.
+
+On the shoot, we create the keys and acquire the WireGuard IP in the standard secret generation. The data is added as a secret to the control plane and to the shootstate. The vpn shoot deployment is extended to include the WireGuard device setup inside the vpn shoot pod network. For certain infrastructures (AWS), we need a re-advertiser to resolve the seed WireGuard endpoint and evaluate whether the IP address changed.
+
+The load balancer exposing the open vpn endpoint is no longer required and therefore removed if this functionality is used.
+
+As we want to slowly expand the usage of the WireGuard solution, we propose to introduce a feature gate for it. Furthermore, since the WireGuard network requires a separate network range, we propose to introduce a new section to the seed settings with two additional flags (enabled & cidr):
+
+```
+apiVersion: core.gardener.cloud/v1beta1
+kind: Seed
+metadata:
+  name: my-seed
+  ...
+spec:
+  ...
+  settings:
+  ...
+    wireguard:
+      enabled: true
+      cidr: 192.168.0.0/22
+```
+
+Last but not least, we propose to introduce an annotation to the shoots to enable/disable the WireGuard tunnel explicitly.
+
+```
+apiVersion: core.gardener.cloud/v1beta1
+kind: Shoot
+metadata:
+  name: my-shoot
+  annotations:
+    alpha.featuregates.shoot.gardener.cloud/wireguard-tunnel: "true"
+  ...
+```
+
+Using this approach, it is easy to switch the solution on and off, i.e. migrate the shoot clusters automatically during ordinary reconciliation.
+
+## High Availability
+
+There is an issue if the node that hosts the WireGuard endpoint fails. The endpoint is migrated to another node however the time required to do this might exceed the budget for downtimes although one could argue that a disruption of less than 30 seconds to 1 minute does not qualify as a downtime and will in almost all cases not noticeable by end users. 
+
+In this case we also assume that TCP connections won't be interrupted - they would just appear to hang. We will confirm this as part of the development and testing effort.
```

Comment:
Why isn't this confirmed already, i.e., how should the reviewers reasonably decide to proceed with this GEP if fundamental questions like this one are not answered yet?

Reply:
Fundamental? We don't have that today as far as I can see. In any case, the proposed solution will be better.

Reply:
Maybe I used the wrong words, sorry for that. I was rather concerned about the sentence 

> We will confirm this as part of the development and testing effort.

Before we decide to proceed or agree with a concrete proposal, we should have all the information requirement to take the decision. I cannot judge the technical relevance of this particular aspect, but if we would hit walls in case we find out later that TCP connections are interrupted (and don't just hang as it's the assumption) then we have a problem...

I would appreciate if you can replace all occurrences of "we need to see this as part of DEV/testing" in this document with results from PoCs, if possible.

Reply:
This is not relevant anymore as we implemented a different proposal.

What I wanted to state is that this solution is most likely better than the existing solution where tcp connections are interrupted, and worst it is as good as the existing solution. As for the potential downtime we will figure this out in a different context where we do experiment with HA setups.